node/core/approval-voting/src/criteria.rs

```diff
@@ -0,0 +1,339 @@
+		let key = session_info.assignment_keys.iter().enumerate()
+			.filter_map(|(i, p)| match keystore.key_pair(p) {
+				Ok(pair) => Some((i as ValidatorIndex, pair)),
+				Err(sc_keystore::Error::PairNotFound(_)) => None,
+				Err(e) => {
+					tracing::warn!(target: LOG_TARGET, "Encountered keystore error: {:?}", e);
+					None
+				}
+			})
+			.next();
+
+		match key {
+			None => return Default::default(),
+			Some(k) => k,
+		}
+	};
+
+	let assignments_key: &sp_application_crypto::sr25519::Pair = assignments_key.as_ref();
+	let assignments_key: &schnorrkel::Keypair = assignments_key.as_ref();
+
+	let mut assignments = HashMap::new();
+
+	// First run `RelayVRFModulo` for each sample.
+	compute_relay_vrf_modulo_assignments(
+		&assignments_key,
+		index,
+		session_info,
+		relay_vrf_story.clone(),
+		leaving_cores.clone(),
+		&mut assignments,
+	);
+
+	// Then run `RelayVRFDelay` once for the whole block.
+	compute_relay_vrf_delay_assignments(
+		&assignments_key,
+		index,
+		session_info,
+		relay_vrf_story,
+		leaving_cores,
+		&mut assignments,
+	);
+
+	assignments
+}
+
+fn compute_relay_vrf_modulo_assignments(
+	assignments_key: &schnorrkel::Keypair,
+	validator_index: ValidatorIndex,
+	session_info: &SessionInfo,
+	relay_vrf_story: RelayVRFStory,
+	leaving_cores: impl IntoIterator<Item = CoreIndex> + Clone,
+	assignments: &mut HashMap<CoreIndex, OurAssignment>,
+) {
+	for rvm_sample in 0..session_info.relay_vrf_modulo_samples {
+		let mut core = Default::default();
+		let maybe_assignment = assignments_key.vrf_sign_extra_after_check(
+			relay_vrf_modulo_transcript(relay_vrf_story.clone(), rvm_sample),
+			|vrf_in_out| {
+				core = relay_vrf_modulo_core(&vrf_in_out, session_info.n_cores);
+				if leaving_cores.clone().into_iter().any(|c| c == core) {
+					Some(assigned_core_transcript(core))
+				} else {
+					None
+				}
+			}
+		);
+
+		if let Some((vrf_in_out, vrf_proof, _)) = maybe_assignment {
+			// Sanity: `core` is always initialized to non-default here, as the closure above
+			// has been executed.
+
+			let cert = AssignmentCert {
+				kind: AssignmentCertKind::RelayVRFModulo { sample: rvm_sample },
+				vrf: (approval_types::VRFOutput(vrf_in_out.to_output()), approval_types::VRFProof(vrf_proof)),
+			};
+
+			// All assignments of type RelayVRFModulo have tranche 0.
+			assignments.entry(core).or_insert(OurAssignment {
+				cert,
+				tranche: 0,
+				validator_index,
+				triggered: false,
+			});
+		}
+	}
+}
+
+fn compute_relay_vrf_delay_assignments(
+	assignments_key: &schnorrkel::Keypair,
+	validator_index: ValidatorIndex,
+	session_info: &SessionInfo,
+	relay_vrf_story: RelayVRFStory,
+	leaving_cores: impl IntoIterator<Item = CoreIndex>,
+	assignments: &mut HashMap<CoreIndex, OurAssignment>,
+) {
+	for core in leaving_cores {
+		let (vrf_in_out, vrf_proof, _) = assignments_key.vrf_sign(
+			relay_vrf_delay_transcript(relay_vrf_story.clone(), core),
+		);
```

Comment:
We could use `vrf_sign_extra_after_check` here too, but it does not give us much really..

Reply:
The goal I had of using `vrf_sign_extra_after_check` elsewhere was to ensure that the VRF proof committed to the assigned core. In this case, the core is part of the initial transcript so I think we don't need it.

I don't mind having every validator assigned to every core with the understanding that most assignments will never be triggered.

Reply:
As you like.  If we've 1000 cores then we'd save 0.025 seconds every 6 seconds (25 microseconds per core) times 1-f where f is this hypothetical tranche cap divided by `num_delay_tranches`.

Reply:
Oh, I see. Well, I'll keep it in mind but I think DB will be a way bigger part of the performance.